---
title: "Ejercicios Módulo 3"
author: "Miguel Arregui Roig"
date: "`r format(Sys.Date(),'%e de %B, %Y')`"
output:
  pdf_document:
    toc: yes
    number_section: yes
    fig_caption: yes
  html_document:
    toc: yes
    toc_depth: 3
    toc_float: yes
    number_sections: yes
    theme: "paper"
# csl: apa.csl
# bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = NULL, cache = TRUE)

```

```{r libraries, include=FALSE}

# Install packages 
# Load packages
# ...

library(knitr)

```

# Ejercicios módulo 3 Análisis de datos Ómicos. Análisis de microarrays

## Methods

### Environment preparation:

Crear un directorio principal, otro directorios **data** para guardar los archivos
*.CEL* y *targets*, y un último directoria llamado **results** para guardar los resultados de análisis de miroarrays.

```{r}
setwd(".")
dir.create("data")
dir.create("results")
dir.create("figures")
```

### Prepare the data for the analysis

```{r TargetsRead}
targets <- read.csv2("./data/targets.csv", header = TRUE, sep = ";")
knitr::kable(targets, booktabs = TRUE, 
             caption = 'Content of the targets file used for the analysis')

```


### Packages installation in R

```{r installPackages, message=FALSE, warning=FALSE, eval=FALSE}
#install.packages("colorspace")
#install.packages("gplots")
#install.packages("ggplot2")
#install.packages("ggrepel")
#install.packages("BiocManager")
#BiocManager::install("oligo")
#BiocManager::install("pd.mogene.2.1.st")
#BiocManager::install("arrayQualityMetrics")
#BiocManager::install("limma")
#BiocManager::install("genefilter")
#BiocManager::install("pvca")
#BiocManager::install("mogene21sttranscriptcluster.db")
#BiocManager::install("annotate")
#BiocManager::install("org.Mm.eg.db")
#BiocManager::install("ReactomePA")

# o comprobar que están acualizados
#if (!require("BiocManager", quietly = TRUE))
 # install.packages("BiocManager")
# BiocManager::install()
```

### Tead the CEL files

Load the package *oligo* in other to read the CEL files. 

```{r}
library(oligo)
celFiles <- list.celfiles("./data", full.names = TRUE)
library(Biobase)
my.targets <- read.AnnotatedDataFrame(file.path("./data", "targets.csv"), 
                                                header = TRUE, row.names = 1, 
                                                sep = ";")
rawData <- read.celfiles(celFiles, phenoData = my.targets)
```


Hemos creado un `EspressionSet`. Podemos cambiar la información del nombre de 
las muestras por uno más comprensible, codificado precuamente en la columna 
*ShortName* del aarchivo *targets*.

```{r}
my.targets@data$ShortName ->rownames(pData(rawData))
colnames(rawData) <- rownames(pData(rawData))
head(rawData)
```


### Quality control of raw data

El siguiente paso es comprobar que los datos crudos guardados presentan suficiente calidad para proceder a la normalización. Para ello disponemos del paquete 
`ArrayQualityMetrics` que realiza diferentes aproximaciones a la calidad de los datos.  Es importante indicar correctamente el archivo de destino de los resultado

```{r QCRaw, message=FALSE, warning=FALSE, eval=FALSE}
require(arrayQualityMetrics)
arrayQualityMetrics(rawData, outdir = file.path("./results", "QCDir.Raw"), force=TRUE)

```


También podemos realizar un PCA análisis con la siguiente función.

```{r}
require(ggplot2)
require(ggrepel)
plotPCA3 <- function (datos, labels, factor, title, scale,colores, size = 1.5, glineas = 0.25) {
  data <- prcomp(t(datos),scale=scale)
  # plot adjustments
  dataDf <- data.frame(data$x)
  Group <- factor
  loads <- round(data$sdev^2/sum(data$sdev^2)*100,1)
  # main plot
  p1 <- ggplot(dataDf,aes(x=PC1, y=PC2)) +
    theme_classic() +
    geom_hline(yintercept = 0, color = "gray70") +
    geom_vline(xintercept = 0, color = "gray70") +
    geom_point(aes(color = Group), alpha = 0.55, size = 3) +
    coord_cartesian(xlim = c(min(data$x[,1])-5,max(data$x[,1])+5)) +
    scale_fill_discrete(name = "Group")
  # avoiding labels superposition
  p1 + geom_text_repel(aes(y = PC2 + 0.25, label = labels),segment.size = 0.25, size = size) + 
    labs(x = c(paste("PC1",loads[1],"%")),y=c(paste("PC2",loads[2],"%"))) +  
    ggtitle(paste("Principal Component Analysis for: ",title,sep=" "))+ 
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_color_manual(values=colores)
  }
```
 Y a continuación mostramos un grtáfico con las dos primeras componentes principales realizadas sobre los datos crudos. 
 
```{r PCARaw, message=FALSE, fig.cap="Visualization of the two first Principal Components for raw data"}
plotPCA3(exprs(rawData), labels = targets$ShortName, factor = targets$Group, 
         title="Raw data", scale = FALSE, size = 3, 
         colores = c("red", "blue", "green", "yellow"))

```
 
If necessary, it is easy to save the plots to a _tiff_ file with the following code:

```{r savePCAraw, echo=TRUE, results='hide'}
tiff("figures/PCA_RawData.tiff", res = 200, width = 4.5, height = 4, units = 'in')
plotPCA3(exprs(rawData), labels = targets$ShortName, factor = targets$Group, 
         title="Raw data", scale = FALSE, size = 2, 
         colores = c("red", "blue", "green", "yellow"))
dev.off()
``` 

También podemos ver la intensidad de la distribución de los arrays realizanzo un boxplot. 

```{r BoxplotRaw, message=FALSE, fig.cap="Boxplot for arrays intensities (Raw Data)"}
boxplot(rawData, cex.axis=0.5, las=2,  which="all", 
         col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
         main="Distribution of raw intensity values")
```

```{r saveIntensRaw, echo=FALSE, results='hide'}
tiff("figures/Intensity_RawData.tiff", res = 200, width = 4, height = 4, units = 'in')
boxplot(rawData, cex.axis=0.5, las=2,  which="all", 
         col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
         main="Distribution of raw intensity values")
dev.off()
```


### Data Normalization

Antes de empezar el análisis de expresión diferencias del array, es necesario hacer que los arrays sean comparables entre ellos y reducir o eliminar la variación entre las muestras que no sean debidas a motivos biológicos. Para esto se realiza el proceso de normalización. 

```{r Normalization}
eset_rma <- rma(rawData)

```

### Quality control of normalized data

```{r QC Normalized, message=FALSE, warning=FALSE, eval=FALSE}
arrayQualityMetrics(eset_rma,  outdir = file.path("./results", "QCDir.Norm"), force=TRUE)

```


Después realizamos un análisis de PCA


```{r PCANorm, message=FALSE, fig.cap="Visualization of first two principal components for normalized data"}
plotPCA3(exprs(eset_rma), labels = targets$ShortName, factor = targets$Group, 
         title="Normalized data", scale = FALSE, size = 3, 
         colores = c("red", "blue", "green", "yellow"))
```

```{r savePCAnorm, echo=FALSE, results='hide'}
tiff("figures/PCA_NormData.tiff", res = 150, width = 5, height = 5, units = 'in')
plotPCA3(exprs(eset_rma), labels = targets$ShortName, factor = targets$Group, 
         title="Normalized data", scale = FALSE, size = 2, 
         colores = c("red", "blue", "green", "yellow"))
dev.off()
```

Y finalmente volvemos a realizar el boxplot, en este caso de los datos normalizados. 

```{r BoxplotNorm, message=FALSE, fig.cap="Distribution of  intensities for normalized data"}
boxplot(eset_rma, cex.axis=0.5, las=2,  which="all", 
         col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
         main="Boxplot for arrays intensity: Normalized Data")
```

```{r saveIntensNorm, echo=FALSE, results='hide'}
tiff("figures/Intensity_NormData.tiff", res = 150, width = 5, height = 5, units = 'in')
boxplot(eset_rma, cex.axis=0.5, las=2,  which="all", 
         col = c(rep("red", 3), rep("blue", 3), rep("green", 3), rep("yellow", 3)),
         main="Boxplot for arrays intensity: Normalized Data")
dev.off()
```


### Batch detection

```{r BatchDetection, message=False, warning=FALSE}
# loas library
require(pvca)
pData(eset_rma) <- targets

# select de threshold
pct_threshold <- 0.6
# select the factors to analyze
batch.factors <- c("Genotype", "Temperature")
# run the analysis
pvcaObj <- pvcaBatchAssess(eset_rma, batch.factors, pct_threshold)

```

```{r plotPVCA, fig.cap="Relative importance of the different factors -genotype, temperature and interaction - affecting gene expression"}

# plot PVCA results
bp <- barplot(pvcaObj$dat, xlab = "Effects", 
              ylab = "Weighted average proportion variance", ylim = c(0, 1.1),
              col = c("mediumorchid"), las = 2, main = "PVCA estimation")
axis(1, at = bp, labels = pvcaObj$label, cex.axis = 0.55, las = 2)
values = pvcaObj$dat
new_values = round(values, 3)
text(bp, pvcaObj$dat, labels = new_values, pos = 3, cex = 0.5)

```

Guardamos el gráfico en la carpeta correspondiente.

```{r savePVCAplot, echo =FALSE, results='hide'}
tiff("figures/PVCAplot.tiff", res = 150, width = 5, height = 5, units = 'in')
bp <- barplot(pvcaObj$dat, xlab = "Effects", 
              ylab = "Weighted average proportion variance", ylim = c(0, 1.1),
              col = c("mediumorchid"), las = 2, main = "PVCA estimation")
axis(1, at = bp, labels = pvcaObj$label, cex.axis = 0.55, las = 2)
values = pvcaObj$dat
new_values = round(values, 3)
text(bp, pvcaObj$dat, labels = new_values, pos = 3, cex = 0.5)
dev.off()

```


### Detecting most variable genes

Graficar la variabilidad media de todos lo genes es util para saber qué porcentaje de genes muestran una variablilidad que se puede atribuir a otras causas más allá de la variabilidad aleatoria. 

```{r SDPlot, fig.cap="Values of standard deviations allong all samples for all genes ordered fron samallest to biggest"}
sds <- apply(exprs(eset_rma), 1, sd)
sdsO <- sort(sds)
plot(1:length(sdsO), sdsO, main = "Distributios of variability for all genes", 
     sub = "Vertical lines represent 90% and 95% percentiles", 
     xlab = "Gene index (from least to most variable)",
     ylab = "Standard deviation")
abline(v = length(sds)*c(0.9, 0.95))

```


Guardamos el gráfico generado

```{r saveSDplot, echo=FALSE, results='hide'}
tiff("figures/SDplot.tiff", res = 150, width = 5, height = 5, units = 'in')
plot(1:length(sdsO), sdsO, main = "Distributios of variability for all genes", 
     sub = "Vertical lines represent 90% and 95% percentiles", 
     xlab = "Gene index (from least to most variable)",
     ylab = "Standard deviation")
abline(v = length(sds)*c(0.9, 0.95))
dev.off()

```


### Filtering least variable genes

A continuación se filtrarán aquellos genes cuya varaibilidad se puede atribuir a la aleaatoriedad, ya que se espera que no tengan una expresión diferencial. Para ellos emplearemos la funación `nsFilter`del paquete de bioconductor `geneFilter`

```{r Filtering1, results='hide', message=FALSE}
require(genefilter)
require(mogene21sttranscriptcluster.db)
annotation(eset_rma) <- "mogene21sttranscriptcluster.db"
filtered <- nsFilter(eset_rma, require.entrez = TRUE, remove.dupEntrez = TRUE, 
                     var.filter = TRUE, var.func = IQR, var.cutoff = 0.75, 
                     filterByQuantile = TRUE, feature.exclude = "^AFFX")


```

```{r FilterResults1, results='hide', echo=FALSE}
names(filtered)
class(filtered)

```
```{r FilterResults2}
print(filtered$filter.log)
eset_filtered <- filtered$eset

```


### Saving normalized and filteres data

```{r SaveData1, results='hide', message=FALSE} 
write.csv(exprs(eset_rma), file = "./results/normalized.Data.csv")
write.csv(exprs(eset_filtered), file = "./results/normalized.Filtered.Data.csv")
save(eset_rma, eset_filtered, file = "./results/normalized.Data.Rda")

```

### Defining the experimental setup: The design matrix

En este modelo vamos a aplicar los modelos lineales para microarrays implementados en el paquete `limma`. 

El primer paso para análisis basados en modelos lineales es crear la matriz de diseño.

```{r LoadSavedData}
if (!exists("eset_filtered")) load (file = "./results/nomarlized.Data.Rda")

```

```{r DesignMatrix, message=FALSE}
require(limma)
designMat <- model.matrix(~ 0 + Group, pData(eset_filtered))
colnames(designMat) <- c("KO.COLD", "KO.RT", "WT.COLD", "WT.RT")
print(designMat)

```


### Defining comparisons with the Contrast Matrix

En este caso, queremos comparar el efecto de gene KO vs WT, de manera independiente para las condiciones RT y COLD. También queremos comparar si hay alguna intereacción entre el gen y la temperatura, para ello tenemos que hacer 3 comparaciones:

```{r, setContrast}
cont.matrix <- makeContrasts(KOvsWT.COLD = KO.COLD-WT.COLD, 
                             KOvsWT.RT = KO.RT-WT.RT, 
                             INT = (KO.COLD-WT.COLD) - (KO.RT -WT.RT),
                             levels = designMat)
print(cont.matrix)

```


### Model estimation and gene selection

```{r, linealmodelfit}
require(limma)
fit <- lmFit(eset_filtered, designMat)
fit.main <- contrasts.fit(fit, cont.matrix)
fit.main <- eBayes(fit.main)
class(fit.main)

```


### Obtaining lists of differentially expressed genes

la función `topTable` del paquete `limma` contiene, para un contraste dado, una lista de genes ordenados de menor a mayor p-valor, por lo que se pueden considerar desde los más a los menos expresados diferencialmente. 

Para la comparación KOvsWT.COLD

```{r topTabs1}
topTab_KOvsWT.COLD <- topTable(fit.main, number = nrow(fit.main), 
                               coef = "KOvsWT.COLD", adjust = "fdr")
head(topTab_KOvsWT.COLD)

```
 
 Para la comparación KOvsWT.RT
 
```{r topTabs2}
topTab_KOvsWT.RT <- topTable(fit.main, number = nrow(fit.main), 
                               coef = "KOvsWT.RT", adjust = "fdr")
head(topTab_KOvsWT.RT)

```
 
 Para la comparación INT
 
```{r topTabs3}
topTab_INT <- topTable(fit.main, number = nrow(fit.main), 
                               coef = "INT", adjust = "fdr")
head(topTab_INT)

```
 
En la primera columna de cada tabla tenemos el ID para cada probeset de Affymetrix.

### Gene Annotation

En este proceso se busca información asociada con los IDs que aparecen en las 
top tables, para darles nombres de Gene Symbol, Entrez Gene identifiers o Gene description. 

Al haber 3 top tables, para simplificar el proceso, se define una función que realice este proceso. 

```{r Gene Annotation, message=FALSE, warning=FALSE}

annotatedTopTable <- function(topTab, anotPackage){
  topTab <- cbind(PROBEID = rownames(topTab), topTab)
  myProbes <- rownames(topTab)
  thePackage <- eval(parse(text = anotPackage))
  geneAnots <- select(thePackage, myProbes, c("SYMBOL", "ENTREZID", "GENENAME"))
  annotatedTopTab <- merge(x = geneAnots, y = topTab, by.x = "PROBEID", 
                           by.y = "PROBEID")
  return(annotatedTopTab)
  }

```


Aplicamos la función a las top tables realizadas

```{r annotatedTopTables}
topAnnotated_KOvsWT.COLD <- annotatedTopTable(topTab_KOvsWT.COLD, anotPackage = "mogene21sttranscriptcluster.db")
topAnnotated_KOvsWT.RT <- annotatedTopTable(topTab_KOvsWT.RT, anotPackage = "mogene21sttranscriptcluster.db")
topAnnotated_INT<- annotatedTopTable(topTab_INT, anotPackage = "mogene21sttranscriptcluster.db")
write.csv(topAnnotated_KOvsWT.COLD, file = "./results/topAnnotated_KOvsWT.COLD.csv")
write.csv(topAnnotated_KOvsWT.RT, file = "./results/topAnnotated_KOvsWT.RT.csv")
write.csv(topAnnotated_INT, file = "./results/topAnnotated_INT.csv")

```

Podemos visualizar una tabla ya anotada

```{r annotatedTop, echo=FALSE}
short <- head(topAnnotated_KOvsWT.COLD[1:5, 1:4])
knitr::kable(
  short, booktabs = TRUE, caption = 'Annotations added to results "topTable" for the comparison "KOvsWT.COLD"'
)

```

### Visualizing differential expression:

Para visualizar expresión diferencial se pueden emplear volcano-plots. 

```{r volcanoPlot, fig.cap="Volcano plot for the comparison between KO and WT in COLD temperature. The namos of the top 4 genes(i.e. the first 4 genes in the topTable) are shown in the plot}
require(mogene21sttranscriptcluster.db)
geneSymbols <- select(mogene21sttranscriptcluster.db, row.names(fit.main), 
                      c("SYMBOL"))
SYMBOLS <- geneSymbols$SYMBOL
volcanoplot(fit.main, coef = 1, highlight = 4, names = SYMBOLS, 
            main = paste("Differentially expressed genes", colnames(cont.matrix)[1], sep = "\n"))
abline(v = c(-1, 1))

```

Guardamos el volcano plot en la carpeta correspondiente

```{r saveVolcanos, echo=FALSE, results='hide'}
tiff("figures/VolcanoPlot.tiff", res = 150, width = 5, height = 5, units = 'in')
volcanoplot(fit.main, coef = 1, highlight = 4, names = SYMBOLS, 
            main = paste("Differentially expressed genes", colnames(cont.matrix)[1], sep = "\n"))
abline(v = c(-1, 1))
dev.off()

pdf("figures/Volcanos.pdf")
for (i in colnames(cont.matrix)){
  volcanoplot(fit.main, coef = i, highlight = 4, names = SYMBOLS, 
              main = paste("Differentially expressed genes", i, sep = "\n"))
  abline(v = c(-1, 1))
}
dev.off()
```

### Multiple Comparisons
